File: packages/moss-repo-indexer-js/src/discover.ts
import { spawnSync } from 'node:child_process'
import fs from 'node:fs'
import path from 'node:path'

import { IndexOptions, resolveIndexOptions } from './types.js'

export function discoverFiles(root: string, options: IndexOptions = {}): string[] {
  const opts = resolveIndexOptions(options)
  const resolvedRoot = path.resolve(root)
  if (!fs.existsSync(resolvedRoot) || !fs.statSync(resolvedRoot).isDirectory()) {
    throw new Error(`Repository root is not a directory: ${resolvedRoot}`)
  }

  const candidates = listCandidates(resolvedRoot, opts)
  return candidates
    .filter((filePath) => isUnderRoot(filePath, resolvedRoot))
    .filter((filePath) => matchesInclude(filePath, resolvedRoot, opts.includeGlobs))
    .filter((filePath) => withinSizeLimit(filePath, opts.maxFileBytes))
    .sort()
}

function listCandidates(root: string, options: ReturnType<typeof resolveIndexOptions>): string[] {
  if (options.respectGitignore && isGitWorkTree(root)) {
    const tracked = gitListFiles(root)
    if (tracked !== null) {
      return tracked
    }
  }
  return walkFilesystem(root, new Set(options.excludeDirs))
}

function isGitWorkTree(root: string): boolean {
  if (!fs.existsSync(path.join(root, '.git'))) {
    return false
  }
  const result = spawnSync('git', ['-C', root, 'rev-parse', '--is-inside-work-tree'], {
    encoding: 'utf8',
  })
  return result.status === 0 && result.stdout.trim() === 'true'
}

function gitListFiles(root: string): string[] | null {
  const version = spawnSync('git', ['--version'], { encoding: 'utf8' })
  if (version.status !== 0) {
    return null
  }
  const result = spawnSync(
    'git',
    ['-C', root, 'ls-files', '-z', '--cached', '--others', '--exclude-standard'],
    { encoding: 'buffer' },
  )
  if (result.status !== 0 || !result.stdout) {
    return null
  }
  const relPaths = result.stdout.toString('utf8').split('\0').filter(Boolean)
  return relPaths
    .map((rel) => path.join(root, rel))
    .filter((filePath) => isIndexableFile(filePath, root))
}

function walkFilesystem(root: string, excludeDirs: Set<string>): string[] {
  const files: string[] = []
  const stack = [root]
  while (stack.length > 0) {
    const current = stack.pop()!
    let entries: fs.Dirent[]
    try {
      entries = fs.readdirSync(current, { withFileTypes: true })
    } catch {
      continue
    }
    for (const entry of entries) {
      const full = path.join(current, entry.name)
      if (entry.isSymbolicLink()) {
        if (isIndexableFile(full, root)) {
          files.push(full)
        }
        continue
      }
      if (entry.isDirectory()) {
        if (!excludeDirs.has(entry.name)) {
          stack.push(full)
        }
        continue
      }
      if (entry.isFile() && isIndexableFile(full, root)) {
        files.push(full)
      }
    }
  }
  return files
}

function isIndexableFile(filePath: string, root: string): boolean {
  try {
    const stat = fs.lstatSync(filePath)
    if (stat.isDirectory()) {
      return false
    }
    if (stat.isSymbolicLink() || stat.isFile()) {
      return isUnderRoot(filePath, root)
    }
    return false
  } catch {
    return false
  }
}

function isUnderRoot(filePath: string, root: string): boolean {
  try {
    const resolved = fs.realpathSync(filePath)
    const resolvedRoot = fs.realpathSync(root)
    const rel = path.relative(resolvedRoot, resolved)
    return rel === '' || (!rel.startsWith('..') && !path.isAbsolute(rel))
  } catch {
    return false
  }
}

function matchesInclude(filePath: string, root: string, includeGlobs: string[]): boolean {
  const rel = toPosix(path.relative(root, filePath))
  if (rel.startsWith('..')) {
    return false
  }
  const name = path.basename(filePath)
  for (const pattern of includeGlobs) {
    if (globMatch(rel, pattern) || globMatch(name, pattern)) {
      return true
    }
    if (pattern.startsWith('**/') && globMatch(name, pattern.slice(3))) {
      return true
    }
  }
  return false
}

function withinSizeLimit(filePath: string, maxFileBytes: number): boolean {
  try {
    return fs.statSync(filePath).size <= maxFileBytes
  } catch {
    return false
  }
}

function toPosix(value: string): string {
  return value.split(path.sep).join('/')
}

/** Minimal glob matcher for include patterns (star and double-star globs). */
function globMatch(value: string, pattern: string): boolean {
  const escaped = pattern
    .replace(/[.+^${}()|[\]\\]/g, '\\$&')
    .replace(/\*\*/g, '{{GS}}')
    .replace(/\*/g, '[^/]*')
    .replace(/{{GS}}/g, '.*')
  return new RegExp(`^${escaped}$`).test(value)
}

File: packages/moss-repo-indexer/src/moss_repo_indexer/discover.py
"""Walk a repository root and yield files to index."""

from __future__ import annotations

import fnmatch
import os
import subprocess
from collections.abc import Iterable
from pathlib import Path

from .types import IndexOptions


def discover_files(root: Path, options: IndexOptions) -> list[Path]:
    """Return absolute paths of files that should be chunked."""
    root = root.expanduser().resolve()
    if not root.is_dir():
        raise FileNotFoundError(f"Repository root is not a directory: {root}")

    candidates = _list_candidates(root, options)
    matched = [
        path
        for path in candidates
        if _is_under_root(path, root) and _matches_include(path, root, options.include_globs)
    ]
    sized = [path for path in matched if _within_size_limit(path, options.max_file_bytes)]
    return sorted(sized)


def _list_candidates(root: Path, options: IndexOptions) -> list[Path]:
    if options.respect_gitignore and _is_git_work_tree(root):
        tracked = _git_list_files(root)
        if tracked is not None:
            return tracked
    return _walk_filesystem(root, set(options.exclude_dirs))


def _is_git_work_tree(root: Path) -> bool:
    if not (root / ".git").exists():
        return False
    result = subprocess.run(
        ["git", "-C", str(root), "rev-parse", "--is-inside-work-tree"],
        capture_output=True,
        text=True,
        check=False,
    )
    return result.returncode == 0 and result.stdout.strip() == "true"


def _git_list_files(root: Path) -> list[Path] | None:
    """List tracked and untracked non-ignored files under root.

    Uses --cached --others --exclude-standard so the candidate set matches a
    normal working tree (not only committed files). Returns None to fall back
    to a filesystem walk when git is unavailable or fails.
    """
    if subprocess.run(["git", "--version"], capture_output=True, check=False).returncode != 0:
        return None
    result = subprocess.run(
        [
            "git",
            "-C",
            str(root),
            "ls-files",
            "-z",
            "--cached",
            "--others",
            "--exclude-standard",
        ],
        capture_output=True,
        check=False,
    )
    if result.returncode != 0:
        return None
    rel_paths = [p for p in result.stdout.decode("utf-8", errors="replace").split("\0") if p]
    files: list[Path] = []
    for rel in rel_paths:
        path = root / rel
        if _is_indexable_file(path, root):
            files.append(path)
    return files


def _walk_filesystem(root: Path, exclude_dirs: set[str]) -> list[Path]:
    files: list[Path] = []
    for dirpath, dirnames, filenames in os.walk(root, followlinks=False):
        current = Path(dirpath)
        dirnames[:] = [
            name
            for name in dirnames
            if name not in exclude_dirs and not (current / name).is_symlink()
        ]
        for name in filenames:
            path = current / name
            if _is_indexable_file(path, root):
                files.append(path)
    return files


def _is_indexable_file(path: Path, root: Path) -> bool:
    if path.is_symlink():
        return path.exists() and _is_under_root(path, root)
    if not path.is_file():
        return False
    return _is_under_root(path, root)


def _is_under_root(path: Path, root: Path) -> bool:
    try:
        path.resolve().relative_to(root.resolve())
        return True
    except (ValueError, OSError):
        return False


def _matches_include(path: Path, root: Path, include_globs: Iterable[str]) -> bool:
    try:
        rel = path.relative_to(root).as_posix()
    except ValueError:
        return False
    name = path.name
    for pattern in include_globs:
        if fnmatch.fnmatch(rel, pattern) or fnmatch.fnmatch(name, pattern):
            return True
        if pattern.startswith("**/") and fnmatch.fnmatch(name, pattern[3:]):
            return True
    return False


def _within_size_limit(path: Path, max_file_bytes: int) -> bool:
    try:
        return path.stat().st_size <= max_file_bytes
    except OSError:
        return False
